src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2775,7 +2775,7 @@ module Crystal
         types = node_types.map do |type|
           type.accept self
           instance_type = type.type.instance_type
-          unless instance_type.implements?(@program.exception)
+          unless instance_type.implements?(@program.exception) || instance_type.module?
```

Comment:
**issue:** This condition still does not work correctly with union types.
For example, the following example should compile, but the union type is considered invalid for `rescue`:
```cr
module Bar; end
begin
rescue Union(Bar, Exception) # Error: (Bar | Exception) is not a subclass of Exception
end
```

Note: The `|` notation in a `rescue` condition does not construct a `Union` type. It's actually an array of individual types which are then treated specially in the exception handler. So `rescue Bar | Exception` is not equivalent to `rescue Union(Bar, Exception)`. It's more similar to `rescue Bar; rescue Exception`.
So the existing specs with `|` notation do not test union types.

Reply:
I think this special handling is super confusing as it corrodes the union syntax. I wonder if this could be solved differently. Maybe we could deconstruct *any* union, regardless of notation.
This is probably not very relevant because complex type unions are not very common in exception handlers.

Reply:
Oh wow okay, in regards to this PR then, if we want to handle this I imagine we could check if the type is a union and if so run the same logic for each of the union types?

This would at least allow it to work until we determine if we want to do some more robust refactoring.

Reply:
BTW: while trying out a few things, the following will catch Ex3 (because the `Ex1|Ex2` alias is casted as `Exception+`):

```crystal
class Ex1 < Exception; end
class Ex2 < Exception; end
class Ex3 < Exception; end

alias Exs = Ex1 | Ex2

begin
  raise Ex3.new
rescue ex : Exs
  p ex
end
```

But an explicit `rescue Ex1 | Ex2` won't catch Ex3, and that's because it's not a proper union.

Reply:
I pushed up new commit(s) to handle your initial example code @straight-shoota. However it does not handle @ysbaddaden's example. But given the latter reproduces on master at the moment it can probably be handled as a follow up at a later point.

I also felt like at this point tho we need a better error message than `is not a subclass of Exception`. So I did a pass on that to make it a bit more clear depending on the context.

Reply:
Yeah, the unification mentioned in https://github.com/crystal-lang/crystal/pull/14553#discussion_r1592776265 is essentially #2661.
The special parsing of union syntax in rescue handlers is a workaround for this issue. But there's nothing we can do when the union type is already unified to the base type (In the example, `Exs` is already an alias to `Exception` and we could not determine the literal declaration was `Ex1 | Ex2`).